{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[]},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["* Select only eukaryotic (superkingdom) proteins (protein evidence, reviewed)\n","* Filter-out sequences shorter than 40 residues\n","* Filter-out sequences having SP (any evidence)\n","* Select only proteins experimentally verified to be localized into: cytosol, nucleus,\n","mitochondrion, plastid, peroxisome, cell membrane\n","\n","OUTPUT:\n","* A TSV file reporting relevant information about the proteins included in the\n","dataset\n","  1. The protein UniProt accession\n","  2. The organism name\n","  3. The Eukaryotic kingdom (Metazoa, Fungi, Plants, Other)\n","  4. The protein length\n","  5. Whether the protein has a transmembrane helix starting in the first 90 residues (true or false)\n","* A FASTA file reporting the protein sequences\n","\n"],"metadata":{"id":"yE3JJNbHRHSE"}},{"cell_type":"markdown","source":["Advanced Search:\n","(taxonomy_id:2759) AND (existence:1) AND (reviewed:true) NOT (ft_signal:*) AND (length:[40 TO *])AND (fragment:false) AND ((cc_scl_term_exp:SL-0091) OR (cc_scl_term_exp:SL-0191) OR (cc_scl_term_exp:SL-0173) OR (cc_scl_term_exp:SL-0209) OR (cc_scl_term_exp:SL-0204) OR (cc_scl_term_exp:SL-0039))\n","\n","\n","---\n","\n","\n","20210 results\n","\n"],"metadata":{"id":"26jS63rVKFY7"}},{"cell_type":"markdown","source":["print only some filds in the API or append only the fields after using the extracted function"],"metadata":{"id":"PcsGDsCU1MdM"}},{"cell_type":"code","execution_count":null,"metadata":{"id":"9jWVYY0IPmsC"},"outputs":[],"source":["#Import necessary packages and modules\n","import requests\n","from requests.adapters import HTTPAdapter, Retry\n","import json\n","import re\n","\n","#Global variables\n","retries = Retry(total=5, backoff_factor=0.25, status_forcelist=[500, 502, 503, 504])\n","session = requests.Session()\n","session.mount(\"https://\", HTTPAdapter(max_retries=retries))"]},{"cell_type":"code","source":["#Define functions to handle API calls and pagination\n","\n","def get_next_link(headers):\n","    if \"Link\" in headers:\n","        # The regular expression is used to extract the next link for pagination\n","        re_next_link = re.compile(r'<(.+)>; rel=\"next\"')\n","        match = re_next_link.match(headers[\"Link\"])\n","        if match:\n","            return match.group(1)\n","\n","def get_batch(batch_url):\n","    while batch_url:\n","        # Run the API call -> get request\n","        response = session.get(batch_url)\n","        # Will raise an error if an error status code is obtained\n","        response.raise_for_status()\n","        # Get the total number of entries in the search\n","        total = response.headers[\"x-total-results\"]\n","        # Yield the response and the total number of entries\n","        yield response, total\n","        # Get the link to the API call for the next data batch\n","        batch_url = get_next_link(response.headers)"],"metadata":{"id":"soA_XLewJsBH"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Advanced search function: (taxonomy_id:2759) AND (existence:1) AND (reviewed:true) NOT (ft_signal:*) AND (length:[40 TO *]) AND ((cc_scl_term_exp:SL-0091) OR (cc_scl_term_exp:SL-0191) OR (cc_scl_term_exp:SL-0173) OR (cc_scl_term_exp:SL-0209) OR (cc_scl_term_exp:SL-0204) OR (cc_scl_term_exp:SL-0039))\n","batch_size = 500\n","url = f\"https://rest.uniprot.org/uniprotkb/search?format=json&query=%28%28taxonomy_id%3A2759%29+AND+%28existence%3A1%29+AND+%28reviewed%3Atrue%29+NOT+%28ft_signal%3A*%29+AND+%28length%3A%5B40+TO+*%5D%29AND+%28fragment%3Afalse%29+AND+%28%28cc_scl_term_exp%3ASL-0091%29+OR+%28cc_scl_term_exp%3ASL-0191%29+OR+%28cc_scl_term_exp%3ASL-0173%29+OR+%28cc_scl_term_exp%3ASL-0209%29+OR+%28cc_scl_term_exp%3ASL-0204%29+OR+%28cc_scl_term_exp%3ASL-0039%29%29%29&size={batch_size}\"\n","\n","# Set the name of the output file, we want TSV output\n","output_file_TSV = \"negative.tsv\"\n","output_file_fasta = \"negative.fasta\"\n","\n","# Run the API call requiring JSON format and build our own TSV file\n","\n","def extract_fields(entry):\n","    # Definition of the organism field\n","    if \"Metazoa\" in entry[\"organism\"][\"lineage\"]:\n","        kingdom = \"Metazoa\"\n","    elif \"Fungi\" in entry[\"organism\"][\"lineage\"]:\n","        kingdom = \"Fungi\"\n","    elif \"Viridiplantae\" in entry[\"organism\"][\"lineage\"]:\n","        kingdom = \"Viridiplantae\"\n","    else:\n","        kingdom = \"Other\"\n","\n","    # Iterate over the features of the entry\n","    for f in entry[\"features\"]:\n","      tr = False\n","      if f[\"type\"] == \"Transmembrane\":\n","        s = f[\"location\"][\"start\"][\"value\"]\n","        if (s<=90) and (\"Helical\" in f[\"description\"]): #re.search(\"Helical\", f[\"description\"])\n","          tr = True\n","          break\n","\n","    return (entry[\"primaryAccession\"], entry[\"organism\"][\"scientificName\"], kingdom, entry[\"sequence\"][\"length\"], tr, entry[\"sequence\"][\"value\"])"],"metadata":{"id":"J_A493KjKCxM"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["def get_dataset(search_url, extract_function, output_file_name_TSV, output_file_name_fasta):\n","    extracted_json = []\n","    n_total= 0\n","    # Run the API call in batches\n","    for batch, total in get_batch(search_url):\n","        # parse the JSON body of the response\n","        batch_json = json.loads(batch.text)\n","        # extract the features\n","        for entry in batch_json[\"results\"]:\n","            n_total += 1\n","            extracted_json.append(extract_function(entry))\n","    print(n_total)\n","\n","    #write the results in an external file using the TSV format\n","    with open(output_file_name_TSV, \"w\") as ofs:\n","        for entry in extracted_json:\n","            fields = entry[:-1]\n","            # Print the fields in TSV format (tab separated)\n","            print(*fields, sep=\"\\t\", file=ofs)\n","        ofs.close\n","\n","    #write the results in an external file using the FASTA format\n","    with open(output_file_name_fasta, \"w\") as ofs:\n","        for entry in extracted_json:\n","            print(\">\"+entry[0]+\"\\n\"+entry[-1], file=ofs)\n","        ofs.close"],"metadata":{"id":"UYitciVhKHbw"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#We call the above function to obtain our dataset\n","get_dataset(url, extract_fields, output_file_TSV, output_file_fasta)"],"metadata":{"id":"U9vMTThXKKhu","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1726233505650,"user_tz":-120,"elapsed":154375,"user":{"displayName":"Denys Carbini","userId":"05111963337097601549"}},"outputId":"6be3e9ec-0dba-4b9e-e2af-575b8651a25a"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["20210\n"]}]},{"cell_type":"code","source":[],"metadata":{"id":"NB-BiYXGYLoJ"},"execution_count":null,"outputs":[]}]}